File: packages/typescript/src/providers/jules-v1alpha.ts
import { ProviderError, classifyHttpError } from "../errors.js";
import { pollingStream } from "../streaming.js";
import type { Artifact, Event, ProviderId, ProviderCapabilities, ReplayToken, Session, SessionOptions, SessionPage, SessionStatus } from "../types.js";
import { ReplayToken as RT } from "../types.js";
import type { SupportsPlanApproval } from "./extensions.js";

const BASE = "https://jules.googleapis.com/v1alpha";

const STATUS_MAP: Record<string, SessionStatus> = {
  QUEUED: "queued",
  PLANNING: "queued",
  AWAITING_PLAN_APPROVAL: "awaiting_input",
  AWAITING_USER_FEEDBACK: "awaiting_input",
  IN_PROGRESS: "in_progress",
  PAUSED: "paused",
  COMPLETED: "completed",
  FAILED: "failed",
  CANCELLED: "cancelled",
};

async function req(
  apiKey: string,
  method: string,
  path: string,
  body?: unknown,
): Promise<Record<string, unknown>> {
  const resp = await fetch(`${BASE}${path}`, {
    method,
    headers: {
      "X-Goog-Api-Key": apiKey,
      "Content-Type": "application/json",
    },
    body: body !== undefined ? JSON.stringify(body) : null,
  });
  const data = (await resp.json().catch(() => ({}))) as Record<string, unknown>;
  if (!resp.ok) {
    const hdrs: Record<string, string> = {};
    resp.headers.forEach((v, k) => { hdrs[k] = v; });
    throw classifyHttpError(resp.status, data, "jules", hdrs);
  }
  return data;
}

export class JulesV1AlphaProvider implements SupportsPlanApproval {
  readonly id: ProviderId = "jules";
  readonly capabilities: ProviderCapabilities = {
    streaming: false,
    cancellation: true,
    listSessions: true,
    sendMessage: true,
    planApproval: true,
    toolResults: false,
    fileUpload: false,
  };

  constructor(private readonly apiKey: string) {}

  private toSession(data: Record<string, unknown>): Session {
    const rawStatus =
      (data["state"] as string | undefined) ??
      (data["status"] as string | undefined) ??
      "QUEUED";
    const status: SessionStatus = STATUS_MAP[rawStatus] ?? "in_progress";
    const artifacts: Artifact[] = [];
    for (const out of (data["outputs"] as Array<Record<string, unknown>> | undefined) ?? []) {
      const pr = out["pullRequest"] as Record<string, unknown> | undefined;
      if (pr) {
        artifacts.push({
          kind: "pull_request",
          provider: "jules",
          ...(pr["url"] !== undefined ? { url: pr["url"] as string } : {}),
          ...(pr["title"] !== undefined ? { title: pr["title"] as string } : {}),
        });
      }
    }
    const nameStr = (data["name"] as string | undefined) ?? "";
    const id = nameStr.includes("/")
      ? nameStr.split("/").pop()!
      : (data["id"] as string | undefined) ?? nameStr;
    return {
      id,
      provider: "jules",
      status,
      createdAt: new Date((data["createTime"] as string | undefined) ?? new Date().toISOString()).toISOString(),
      updatedAt: new Date((data["updateTime"] as string | undefined) ?? new Date().toISOString()).toISOString(),
      artifacts,
      replayToken: RT._encode({}),
    };
  }

  async createSession(opts: SessionOptions): Promise<Session> {
    const po = opts.providerOptions ?? {};
    const body: Record<string, unknown> = { prompt: opts.prompt };
    if (opts.title) body["title"] = opts.title;
    if (
      opts.source &&
      (opts.source.kind === "github" || opts.source.kind === "gitlab" || opts.source.kind === "bitbucket")
    ) {
      const rest = opts.source.uri.split("://")[1]!;
      const [repoPart, uriB] = rest.split("@", 2) as [string, string | undefined];
      const slashIdx = repoPart.indexOf("/");
      const owner = repoPart.slice(0, slashIdx);
      const repo = repoPart.slice(slashIdx + 1);
      const sourceName = `sources/github/${owner}/${repo}`;
      let branch: string | undefined = opts.source.branch ?? uriB;
      if (!branch) {
        // Jules requires startingBranch on every request — fetch the default
        const srcData = await req(this.apiKey, "GET", `/${sourceName}`);
        const githubRepo = srcData["githubRepo"] as Record<string, unknown> | undefined;
        const defaultBranch = githubRepo?.["defaultBranch"] as Record<string, unknown> | undefined;
        branch = defaultBranch?.["displayName"] as string | undefined;
      }
      body["sourceContext"] = {
        source: sourceName,
        githubRepoContext: { startingBranch: branch },
      };
    }
    if (po["require_plan_approval"]) body["requirePlanApproval"] = true;
    if (po["auto_create_pr"]) body["automationMode"] = "AUTO_CREATE_PR";
    const data = await req(this.apiKey, "POST", "/sessions", body);
    return this.toSession(data);
  }

  async getSession(sessionId: string, _replayToken?: ReplayToken): Promise<Session> {
    const data = await req(this.apiKey, "GET", `/sessions/${sessionId}`);
    return this.toSession(data);
  }

  async *streamEvents(
    sessionId: string,
    replayToken?: ReplayToken,
    sinceSeq = 0,
  ): AsyncIterableIterator<Event> {
    const seenIds = new Set<string>();
    const apiKey = this.apiKey;

    const fetchSession = () => this.getSession(sessionId, replayToken);

    const fetchNewEvents = async (
      cursor: string | undefined,
    ): Promise<[Event[], string | undefined]> => {
      const params = new URLSearchParams({ pageSize: "100" });
      if (cursor) params.set("pageToken", cursor);
      let data: Record<string, unknown>;
      try {
        data = await req(apiKey, "GET", `/sessions/${sessionId}/activities?${params}`);
      } catch (err) {
        // Jules returns 404 on /activities while session is still QUEUED
        if (err instanceof ProviderError && err.kind === "not_found") {
          return [[], cursor];
        }
        throw err;
      }
      const activities = (data["activities"] as Array<Record<string, unknown>> | undefined) ?? [];
      const nextToken = data["nextPageToken"] as string | undefined;
      const events: Event[] = [];
      for (const act of activities) {
        const actId = (act["name"] as string | undefined) ?? "";
        if (seenIds.has(actId)) continue;
        seenIds.add(actId);
        const ev = activityToEvent(act, sessionId);
        if (ev) events.push(ev);
      }
      return [events, nextToken];
    };

    yield* pollingStream({
      provider: "jules",
      sessionId,
      fetchSession,
      fetchNewEvents,
      pollBase: 2000,
      pollCap: 15000,
    });
  }

  async cancelSession(sessionId: string, _replayToken?: ReplayToken): Promise<void> {
    await req(this.apiKey, "DELETE", `/sessions/${sessionId}`);
  }

  async sendMessage(sessionId: string, message: string, _replayToken?: ReplayToken): Promise<void> {
    await req(this.apiKey, "POST", `/sessions/${sessionId}:sendMessage`, { prompt: message });
  }

  async listSessions(limit = 20, cursor?: string): Promise<SessionPage> {
    const params = new URLSearchParams({ pageSize: String(limit) });
    if (cursor) params.set("pageToken", cursor);
    const data = await req(this.apiKey, "GET", `/sessions?${params}`);
    const sessions = (
      (data["sessions"] as Array<Record<string, unknown>> | undefined) ?? []
    ).map((s) => this.toSession(s));
    const nextCursor = data["nextPageToken"] as string | undefined;
    return {
      sessions,
      ...(nextCursor !== undefined ? { nextCursor } : {}),
      provider: "jules",
    };
  }

  async approvePlan(sessionId: string, _replayToken?: ReplayToken): Promise<void> {
    await req(this.apiKey, "POST", `/sessions/${sessionId}:approvePlan`, {});
  }
}

function activityToEvent(act: Record<string, unknown>, sessionId: string): Event | null {
  const tsStr = (act["createTime"] as string | undefined) ?? new Date().toISOString();
  const ts = new Date(tsStr).toISOString();

  // Activity type is indicated by which field is present — there is no "kind" field.
  if ("agentMessaged" in act) {
    const am = act["agentMessaged"] as Record<string, unknown> | undefined;
    return {
      kind: "message",
      sessionId,
      provider: "jules",
      ts,
      seq: 0,
      data: { text: (am?.["agentMessage"] as string | undefined) ?? "" },
      raw: act,
    };
  }
  if ("progressUpdated" in act) {
    const pu = act["progressUpdated"] as Record<string, unknown> | undefined;
    return {
      kind: "tool_activity",
      sessionId,
      provider: "jules",
      ts,
      seq: 0,
      data: {
        tool: (pu?.["tool"] as string | undefined) ?? "",
        summary: (pu?.["summary"] as string | undefined) ?? "",
        title: (pu?.["title"] as string | undefined) ?? "",
      },
      raw: act,
    };
  }
  if ("planGenerated" in act) {
    const pg = act["planGenerated"] as Record<string, unknown> | undefined;
    const plan = pg?.["plan"] as Record<string, unknown> | undefined;
    return {
      kind: "provider_event",
      sessionId,
      provider: "jules",
      ts,
      seq: 0,
      data: {
        type: "plan_proposed",
        steps: (plan?.["steps"] as unknown[] | undefined) ?? [],
      },
      raw: act,
    };
  }
  if ("userMessaged" in act) {
    const um = act["userMessaged"] as Record<string, unknown> | undefined;
    return {
      kind: "provider_event",
      sessionId,
      provider: "jules",
      ts,
      seq: 0,
      data: {
        type: "user_message",
        text: (um?.["userMessage"] as string | undefined) ?? "",
      },
      raw: act,
    };
  }
  // passthrough for planApproved, sessionCompleted, and anything else
  const actType =
    Object.keys(act).find(
      (k) => !["name", "createTime", "originator", "id", "artifacts"].includes(k),
    ) ?? "unknown";
  return {
    kind: "provider_event",
    sessionId,
    provider: "jules",
    ts,
    seq: 0,
    data: { type: actType, raw: act },
    raw: act,
  };
}

File: packages/python/src/omniharness/providers/_jules_v1alpha.py
from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from typing import Any, AsyncIterator, Optional

import httpx

from .._streaming import polling_stream
from ..errors import ProviderError, UnsupportedCapability, classify_http_error
from ..types import (
    Artifact,
    Event,
    ProviderId,
    ProviderCapabilities,
    ReplayToken,
    Session,
    SessionOptions,
    SessionPage,
    SessionStatus,
    Source,
)

_BASE = "https://jules.googleapis.com/v1alpha"

_STATUS_MAP: dict[str, SessionStatus] = {
    "QUEUED": "queued",
    "PLANNING": "queued",
    "AWAITING_PLAN_APPROVAL": "awaiting_input",
    "AWAITING_USER_FEEDBACK": "awaiting_input",
    "IN_PROGRESS": "in_progress",
    "PAUSED": "paused",
    "COMPLETED": "completed",
    "FAILED": "failed",
    "CANCELLED": "cancelled",
}


def _parse_dt(s: str) -> datetime:
    return datetime.fromisoformat(s.replace("Z", "+00:00"))


class JulesV1AlphaProvider:
    id: ProviderId = "jules"
    capabilities = ProviderCapabilities(
        streaming=False,
        cancellation=True,
        list_sessions=True,
        send_message=True,
        plan_approval=True,
        tool_results=False,
        file_upload=False,
    )

    def __init__(self, api_key: str) -> None:
        self._http = httpx.AsyncClient(
            base_url=_BASE,
            headers={"X-Goog-Api-Key": api_key},
            timeout=30.0,
        )

    async def _req(
        self, method: str, path: str, **kwargs: Any
    ) -> httpx.Response:
        resp = await self._http.request(method, path, **kwargs)
        if resp.status_code >= 400:
            raise classify_http_error(
                resp.status_code, resp.json() if resp.content else {}, "jules", dict(resp.headers)
            )
        return resp

    def _to_session(self, data: dict[str, Any]) -> Session:
        raw_status = data.get("state", data.get("status", "QUEUED"))
        status = _STATUS_MAP.get(raw_status, "in_progress")
        artifacts: list[Artifact] = []
        for out in data.get("outputs", []):
            if pr := out.get("pullRequest"):
                artifacts.append(
                    Artifact(
                        kind="pull_request",
                        provider="jules",
                        url=pr.get("url"),
                        title=pr.get("title"),
                    )
                )
        token = ReplayToken._encode({})
        return Session(
            id=data["name"].split("/")[-1] if "/" in data.get("name", "") else data["name"],
            provider="jules",
            status=status,
            created_at=_parse_dt(data.get("createTime", datetime.now(timezone.utc).isoformat())),
            updated_at=_parse_dt(data.get("updateTime", datetime.now(timezone.utc).isoformat())),
            artifacts=artifacts,
            replay_token=token,
        )

    async def create_session(self, opts: SessionOptions) -> Session:
        po = opts.provider_options or {}
        body: dict[str, Any] = {"prompt": opts.prompt}
        if opts.title:
            body["title"] = opts.title
        if opts.source and opts.source.kind in ("github", "gitlab", "bitbucket"):
            parts = opts.source.uri.split("://", 1)[1].split("@", 1)
            repo_path = parts[0]
            branch = opts.source.branch or (parts[1] if len(parts) > 1 else None)
            owner, _, repo = repo_path.partition("/")
            source_name = f"sources/github/{owner}/{repo}"
            if not branch:
                src_resp = await self._req("GET", f"/{source_name}")
                src_data = src_resp.json()
                branch = (
                    src_data.get("githubRepo", {})
                    .get("defaultBranch", {})
                    .get("displayName")
                )
            body["sourceContext"] = {
                "source": source_name,
                "githubRepoContext": {"startingBranch": branch},
            }
        if po.get("require_plan_approval"):
            body["requirePlanApproval"] = True
        automation = po.get("auto_create_pr")
        if automation:
            body["automationMode"] = "AUTO_CREATE_PR"
        resp = await self._req("POST", "/sessions", json=body)
        return self._to_session(resp.json())

    async def get_session(
        self,
        session_id: str,
        replay_token: Optional[ReplayToken] = None,
    ) -> Session:
        resp = await self._req("GET", f"/sessions/{session_id}")
        return self._to_session(resp.json())

    async def stream_events(
        self,
        session_id: str,
        replay_token: Optional[ReplayToken] = None,
        since_seq: int = 0,
    ) -> AsyncIterator[Event]:
        seen_activity_ids: set[str] = set()
        page_token: Optional[str] = None

        async def fetch_session() -> Session:
            return await self.get_session(session_id, replay_token)

        async def fetch_new_events(cursor: Optional[str]) -> tuple[list[Event], Optional[str]]:
            nonlocal page_token
            params: dict[str, Any] = {"pageSize": 100}
            if cursor:
                params["pageToken"] = cursor
            try:
                resp = await self._req("GET", f"/sessions/{session_id}/activities", params=params)
            except ProviderError as exc:
                if exc.kind == "not_found":
                    return [], cursor
                raise
            data = resp.json()
            activities = data.get("activities", [])
            next_token: Optional[str] = data.get("nextPageToken")
            events: list[Event] = []
            for act in activities:
                act_id = act.get("name", "")
                if act_id in seen_activity_ids:
                    continue
                seen_activity_ids.add(act_id)
                ev = _activity_to_event(act, session_id)
                if ev:
                    events.append(ev)
            return events, next_token

        async for ev in polling_stream(
            provider="jules",
            session_id=session_id,
            fetch_session=fetch_session,
            fetch_new_events=fetch_new_events,
            poll_base=2.0,
            poll_cap=15.0,
        ):
            yield ev

    async def cancel_session(
        self,
        session_id: str,
        replay_token: Optional[ReplayToken] = None,
    ) -> None:
        await self._req("DELETE", f"/sessions/{session_id}")

    async def send_message(
        self,
        session_id: str,
        message: str,
        replay_token: Optional[ReplayToken] = None,
    ) -> None:
        await self._req("POST", f"/sessions/{session_id}:sendMessage", json={"prompt": message})

    async def list_sessions(
        self,
        limit: int = 20,
        cursor: Optional[str] = None,
    ) -> SessionPage:
        params: dict[str, Any] = {"pageSize": limit}
        if cursor:
            params["pageToken"] = cursor
        resp = await self._req("GET", "/sessions", params=params)
        data = resp.json()
        sessions = [self._to_session(s) for s in data.get("sessions", [])]
        return SessionPage(
            sessions=sessions,
            next_cursor=data.get("nextPageToken"),
            provider="jules",
        )

    # Extension: SupportsPlanApproval
    async def approve_plan(
        self,
        session_id: str,
        replay_token: Optional[ReplayToken] = None,
    ) -> None:
        await self._req("POST", f"/sessions/{session_id}:approvePlan", json={})


def _activity_to_event(act: dict[str, Any], session_id: str) -> Optional[Event]:
    ts_str = act.get("createTime", datetime.now(timezone.utc).isoformat())
    ts = _parse_dt(ts_str)

    if "agentMessaged" in act:
        return Event(
            kind="message",
            session_id=session_id,
            provider="jules",
            ts=ts,
            seq=0,
            data={"text": act["agentMessaged"].get("agentMessage", "")},
            raw=act,
        )
    if "progressUpdated" in act:
        tool = act["progressUpdated"].get("tool", "")
        summary = act["progressUpdated"].get("summary", "")
        return Event(
            kind="tool_activity",
            session_id=session_id,
            provider="jules",
            ts=ts,
            seq=0,
            data={"tool": tool, "summary": summary},
            raw=act,
        )
    if "planGenerated" in act:
        return Event(
            kind="provider_event",
            session_id=session_id,
            provider="jules",
            ts=ts,
            seq=0,
            data={"type": "plan_proposed", "steps": act["planGenerated"].get("steps", [])},
            raw=act,
        )
    if "userMessaged" in act:
        return Event(
            kind="provider_event",
            session_id=session_id,
            provider="jules",
            ts=ts,
            seq=0,
            data={"type": "user_message", "text": act["userMessaged"].get("userMessage", "")},
            raw=act,
        )
    # anything else → provider_event passthrough
    act_type = next((k for k in act if k not in ("name", "createTime", "originator", "id")), "unknown")
    return Event(
        kind="provider_event",
        session_id=session_id,
        provider="jules",
        ts=ts,
        seq=0,
        data={"type": act_type, "raw": act},
        raw=act,
    )
